src/rsa/key_generation.rs

```diff
@@ -0,0 +1,148 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! RSA key generation
+
+use super::small_primes::SMALL_PRIMES;
+use core::marker::PhantomData;
+use std::vec::Vec;
+
+use crate::{
+    limb::{self, LimbMask},
+    error, rand,
+    rsa::bigint,
+};
+
+/// Performs the Miller-Rabin primality test
+///
+/// Implementation of the algorithm as it
+/// is described in FIPS 186-4 C.3.1
+fn miller_rabin_test(
+    w: &[u64],
+    iterations: u32,
+    rng: &rand::SecureRandom,
+) -> Result<bool, error::Unspecified> {
+    // Step 1.
+    let mut w_m1 = vec![0; w.len()];
+    limb::limbs_sub_limb(&mut w_m1, w, 1);
+    let a = limb::limbs_count_low_zero_bits(&w_m1);
+
+    // Step 2.
+    let mut m = vec![0; w_m1.len()];
+    limb::limbs_copy(&mut m, &w_m1);
+    limb::limbs_rshift(&mut m, a);
+
+    // Step 3.
+    let _wlen = w.len();
+
+    // preparation
+    let boxed_w = bigint::BoxedLimbs::<()>::minimal_width_from_unpadded(&w);
+    let w_modulus = bigint::Modulus::from_boxed_limbs(boxed_w)?.0;
+    let boxed_m = bigint::BoxedLimbs::<()>::minimal_width_from_unpadded(&m);
+    let m_exponent = bigint::PrivateExponent {
+        limbs: boxed_m,
+    };
+    // Step 4.
+    let mut b: Vec<limb::Limb> = vec![0; w.len()];
+    for _ in 0 .. iterations {
+        loop {
+            // Step 4.1
+            unsafe {
+                let b_mut_ptr = (&mut b[..]).as_mut_ptr() as * mut u8;
+                let b_mut_slice = core::slice::from_raw_parts_mut(b_mut_ptr, b.len() * 8);
+                rng.fill(b_mut_slice)?;
+            }
+            // Step 4.2
+            if limb::limbs_less_than_limb_constant_time(&b, 2) == LimbMask::False
+                    && limb::limbs_less_than_limbs_consttime(&b, &w_m1) == LimbMask::True
+            {
+                break;
+            }
+        }
+        // Step 4.3
+        let boxed_b = bigint::BoxedLimbs::minimal_width_from_unpadded(&b);
+        let b_elem = bigint::Elem {
+            limbs: boxed_b,
+            encoding: PhantomData,
+        };
+        let mut z_elem = bigint::elem_exp_consttime(b_elem, &m_exponent, &w_modulus)?;
+        // Step 4.4
+        if z_elem.is_one() || limb::limbs_equal_limbs_consttime(&z_elem.limbs, &w_m1) == LimbMask::True {
+            continue;
+        }
+        // Step 4.5
+        for _ in 1 .. a {
+            // Step 4.5.1
+            // TODO figure out how to make this compile
+            //z_elem = bigint::elem_squared(z_elem, &w_modulus.as_partial());
```

Comment:
@briansmith do you know how this can be made compile? Right now I'm getting sth like:

```error[E0308]: mismatched types                                                                                         
  --> src/rsa/key_generation.rs:87:22                                                                                  
   |                                                                                                                   
87 |             z_elem = bigint::elem_squared(z_elem, &w_modulus.as_partial());                                       
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `arithmetic::montgomery::Unencoded`, found enum `arithmetic::montgomery::RInverse`
   |                                                                                                                   
   = note: expected type `rsa::bigint::Elem<_, arithmetic::montgomery::Unencoded>`                                     
              found type `rsa::bigint::Elem<_, arithmetic::montgomery::RInverse>` 

```

Reply:
For now: multiply `z_elem` by `w_modulus.oneRR` before the loop.

Later: Refactor `elem_exp_consttime` so that it returns an `Elem<R>` instead of `Elem<Unencoded>` by moving the `into_unencoded()` to the caller.

Reply:
When I put this before the loop,
```Rust
        let z_elem_u = bigint::elem_exp_consttime(b_elem, &m_exponent, &w_modulus)? ;
        let mut z_elem = bigint::elem_mul(z_elem_u, w_modulus.oneRR, w_modulus);
```
and I do
```Rust
            z_elem = bigint::elem_squared(z_elem, &w_modulus.as_partial());
```
it gives me:
```
error[E0308]: mismatched types                                                                                         
  --> src/rsa/bigint/key_generation.rs:82:22                                                                           
   |                                                                                                                   
82 |             z_elem = bigint::elem_squared(z_elem, &w_modulus.as_partial());                                       
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `arithmetic::montgomery::Unencoded`, found enum `arithmetic::montgomery::RInverse`
   |                                                                                                                   
   = note: expected type `rsa::bigint::Elem<_, arithmetic::montgomery::Unencoded>`                                     
              found type `rsa::bigint::Elem<_, arithmetic::montgomery::RInverse>`
```

if instead I do
```Rust
z_elem = bigint::elem_mul(&z_elem, z_elem.clone(), &w_modulus);
```

I get a type mismatch as well.

```
error[E0308]: mismatched types                                                                                         
  --> src/rsa/bigint/key_generation.rs:83:22                                                                           
   |                                                                                                                   
83 |             z_elem = bigint::elem_mul(&z_elem, z_elem.clone(), &w_modulus);                                       
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `arithmetic::montgomery::Unencoded`, found enum `arithmetic::montgomery::RInverse`
   |                                                                                                                   
   = note: expected type `rsa::bigint::Elem<_, arithmetic::montgomery::Unencoded>`                                     
              found type `rsa::bigint::Elem<_, arithmetic::montgomery::RInverse>` 
```